test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+      consoleSpy.mockImplementation(() => {});
+      const codeGenSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      auth = new Auth({} as any);
+      auth.login();
+      expect(consoleSpy).toHaveBeenCalled();
+      expect(codeGenSpy).toHaveBeenCalledTimes(0);
+
+      consoleSpy.mockRestore();
+      codeGenSpy.mockRestore();
+    });
+
+    it('opens a window to the wally api login page with clientId and state', () => {
+      const state = 'mississippi';
+      const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
+
+      jest
+        .spyOn(Auth.prototype as any, 'generateStateCode')
+        .mockImplementation(() => state);
+
+      auth.login();
+
+      expect(storageSpy).toHaveBeenCalledWith(
+        `wally:${clientId}:state:token`,
+        state
+      );
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}`,
+        '_blank'
+      );
+
+      storageSpy.mockRestore();
+    });
+
+    it('passes redirectUrl and email when set', () => {
+      const state = 'mississippi';
+      const email = 'joe@smith.net';
+      const redirectURL = 'http://salmonofcapistrano.com';
+
+      const genCodeSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      genCodeSpy.mockImplementation(() => state);
+
+      auth = new Auth({
+        clientId,
+        host,
+        redirectURL,
+        messenger: new Messenger({}),
+      } as any);
+      auth.login(email);
+
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}&redirectUrl=${encodeURIComponent(
+          redirectURL
+        )}&email=${encodeURIComponent(email)}`,
+        '_blank'
+      );
+      genCodeSpy.mockRestore();
+    });
+
+    describe('returned promise', () => {
+      it('attaches message listeners', () => {
+        const emitterSpy = jest.spyOn(Messenger.prototype, 'addListener');
+        const workerSpy = jest.spyOn(Messenger.prototype, 'onWorkerMessage');
+        auth.login();
+        expect(emitterSpy).toHaveBeenCalled();
+        expect(workerSpy).toHaveBeenCalledTimes(2);
+        expect(workerSpy.mock.calls[0][0]).toBe(WorkerMessage.LOGIN_SUCCESS);
+        expect(workerSpy.mock.calls[1][0]).toBe(WorkerMessage.LOGIN_FAILURE);
+
+        emitterSpy.mockRestore();
+        workerSpy.mockRestore();
+      });
+
+      it('resolves on the accounts changed emitter message', async () => {
+        setTimeout(() => {
+          auth['messenger'].emit(EmitterMessage.ACCOUNTS_CHANGED, '0x1234');
+        }, 0);
+
+        await auth.login();
+      });
+
+      it('resolves on the login success worker message when token', async () => {
+        const tokenSpy = jest.spyOn(Auth.prototype, 'getToken');
+        tokenSpy.mockImplementation(() => 'asdf');
+
+        setTimeout(() => {
+          auth['messenger'].handleWorkerMessage(WorkerMessage.LOGIN_SUCCESS);
+        }, 0);
+
+        await auth.login();
```

Comment:
Q: Is the token returned here? If so, since `getToken` is mocked, a stronger test would just be checking that the return value is the mocked return value. Generally, testing for `toHaveBeenCalled` requires implementation details and can produce more fragile tests.

Reply:
`login()` returns a `Promise<void>` - I'm not sure if returning the token would be a good/bad idea? Should the consumer of the API get access to it, or is it better kept internal?

otherwise, similar answer to above, i'm mainly testing that the promise is resolved - i can remove the `toHaveBeenCalled`, not super necessary to the test